onnxruntime/core/framework/sparse_tensor.cc

```diff
@@ -3,31 +3,441 @@
+  ORT_ENFORCE(allocator_ != nullptr, "This method should follow a call to constructor that supplies the allocator");
+  ORT_THROW_IF_ERROR(ValidateBlockSparseShapes(values_shape, index_shape));
+  if (values_shape.Size() > 0) {
+    const auto data_size = values_shape.Size() * ml_data_type_->Size();
+    const auto index_size = index_shape.Size() * sizeof(int32_t);
+    const auto required_buffer_size = CalculateRequiredBufferSize(gsl::narrow<int64_t>(data_size),
+                                                                  gsl::narrow<int64_t>(index_size));
+    ORT_THROW_IF_ERROR(AllocateBuffer(required_buffer_size, static_cast<size_t>(data_size / ml_data_type_->Size())));
+  }
+  values_ = Tensor(DataType(), values_shape, p_data_, Location());
+  format_data_.resize(1);
+  format_data_[0] = Tensor(DataTypeImpl::GetType<int32_t>(), index_shape, IndicesStart(values_.SizeInBytes()), Location());
+  format_ = SparseFormat::kBlockSparse;
+  return BlockSparseMutator(values_, format_data_[0]);
+}
+
+Status SparseTensor::Copy(const DataTransferManager& data_transfer_manager, int exec_q_id, SparseTensor& dst_tensor) const {
+  const IDataTransfer* data_transfer = data_transfer_manager.GetDataTransfer(Location().device,
+                                                                             dst_tensor.Location().device);
+  ORT_RETURN_IF_NOT(data_transfer != nullptr, "Unable to find a data transfer for copying from device type: ",
+                    Location().device.Type(), " to device type: ", dst_tensor.Location().device.Type());
+
+  return Copy(*data_transfer, dst_tensor, exec_q_id);
+}
+
+Status SparseTensor::Copy(const IDataTransfer& data_transfer, SparseTensor& dst_tensor, int exec_q_id) const {
+  // Do not copy same destination
+  if (this == &dst_tensor) {
+    return Status::OK();
+  }
+
+  ORT_RETURN_IF_NOT(format_ != SparseFormat::kUndefined, "This instance should not be empty");
+  const bool is_string = IsDataTypeString();
+
+  ORT_RETURN_IF_NOT(dst_tensor.Format() == SparseFormat::kUndefined, "Destination should be empty");
+  ORT_RETURN_IF_NOT(dst_tensor.allocator_ != nullptr, "Destination must have a CPU allocator set");
+  ORT_RETURN_IF_NOT((!is_string || dst_tensor.Location().device.Type() == OrtDevice::CPU),
+                    "X-device copy of strings not supported");
+  ORT_RETURN_IF_NOT(dst_tensor.DataType() == DataType(), "Src and Dst must be of the same type");
+  ORT_RETURN_IF_NOT(dst_tensor.dense_shape_.Size() == dense_shape_.Size(), "Must have the same shape");
+
+  const auto required_buffer_size = RequiredAllocationSize();
+  SparseTensor result(DataType(), Shape(), dst_tensor.allocator_);
+  ORT_RETURN_IF_ERROR(result.AllocateBuffer(required_buffer_size, NumValues()));
+
+  const auto values_bytes = values_.SizeInBytes();
+  auto* const dst_index_start = reinterpret_cast<int8_t*>(result.IndicesStart(values_bytes));
+  result.format_data_.resize(format_data_.size());
+  int64_t index_bytes = 0;
```

Comment:
>index_bytes [](http://example.com/codeflow?start=10&length=11)

would it be appropriate to make this SafeInt?

Reply:
Sure. However, this is a copy and if the src data is sound, then we should have no issues.